RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc

```diff
@@ -1,78 +1,106 @@
+  max_NormalizedChi2_ (iConfig.getParameter<double> ("MaxNormalizedChi2")),
+  max_PtDifference_ (iConfig.getParameter<double> ("MaxPtDifference")),
+  min_Nhits_ (iConfig.getParameter<int> ("MinNhits")),
+  min_NmuonHits_ (iConfig.getParameter<int> ("MinNmuonHits"))  
+{
+   LogTrace("Muon|RecoMuon|HLTMuonL2SelectorForL3IO")<<"constructor called";
+   produces<reco::TrackCollection>();
 }
-  
+
 /// destructor
 HLTMuonL2SelectorForL3IO::~HLTMuonL2SelectorForL3IO(){}
 
 /// create collection of L2 muons not already reconstructed as L3 muons
 void HLTMuonL2SelectorForL3IO::produce(edm::Event& iEvent, const edm::EventSetup& iSetup){
-	const std::string metname = "Muon|RecoMuon|HLTMuonL2SelectorForL3IO";
-  
-//  IN:
-	edm::Handle<reco::TrackCollection> l2TrackCol;
-	iEvent.getByToken(l2Src_, l2TrackCol);
-
-	edm::Handle<reco::TrackCollection> l3TrackCol;
-	iEvent.getByToken(l3OISrc_, l3TrackCol);
+  const std::string metname = "Muon|RecoMuon|HLTMuonL2SelectorForL3IO";
+	
+  //  IN:
+  edm::Handle<reco::TrackCollection> l2muonH;
+  iEvent.getByToken(l2Src_, l2muonH);
 
-//	OUT:
-	auto result = std::make_unique<std::vector<reco::Track>>();
+  edm::Handle<reco::RecoChargedCandidateCollection> l3muonH;
+  iEvent.getByToken(l3OISrc_,l3muonH);
 
-	auto const& l2Tracks = *l2TrackCol.product();
-	auto const& l3Tracks = *l3TrackCol.product();
-	for (auto&& l2 : l2Tracks){
-		bool l2found=false;
-		for (auto&& l3 : l3Tracks){
-			if (useOuterHitPosition_){
-				// If x,y are within 0.5cm of each other: L2 has been found already
-				// z found to change beyond this (perhaps a refitting issue?)
-				if (std::abs(l2.outerPosition().X()-l3.outerPosition().X()) < xDiffMax_ &&
-					std::abs(l2.outerPosition().Y()-l3.outerPosition().Y()) < yDiffMax_ &&
-					std::abs(l2.outerPosition().Z()-l3.outerPosition().Z()) < zDiffMax_){
-					l2found=true;
-				}
-			}
-			else{
-				if (deltaR(l3,l2)<dRDiffMax_) l2found=true;
-			}
-		}
-		if (!l2found) result->push_back(l2);
+  // Read Links collection:
+  edm::Handle<reco::MuonTrackLinksCollection> links;
+  iEvent.getByToken(l3linkToken_, links);  
+  
+  //	OUT:
+  std::unique_ptr<reco::TrackCollection > result(new reco::TrackCollection());
+    
+  for (unsigned int il2=0; il2 != l2muonH->size(); ++il2){
+    reco::TrackRef l2muRef(l2muonH, il2);
+    bool re_do_this_L2=true;
+    
+    for (unsigned int il3=0; il3 != l3muonH->size(); ++il3){
+      reco::RecoChargedCandidateRef cand(l3muonH,il3);
+      reco::TrackRef tk = cand->track();
+      
+      bool useThisLink = false;
+      for (unsigned int l(0); l<links->size() && !useThisLink; ++l){
+	const reco::MuonTrackLinks* link = &links->at(l);
+	
+	// Check if the L3 link matches the L3 candidate
+	const reco::Track& globalTrack = *link->globalTrack();
+	float dR2 = deltaR2(tk->eta(),tk->phi(),globalTrack.eta(),globalTrack.phi());
+	float dPt = std::abs(tk->pt() - globalTrack.pt())/tk->pt();
+	if (dR2 < 0.02*0.02 and dPt < 0.001) {
```

Comment:
@folguera - better to write this as 
if ( dR2< 0.02*0.02 and std::abs(tk->pt() - globalTrack.pt()) < 0.001*tk->pt() ) {....